### DISCUSSION TITLE
Get list of published topics, not just subscription patterns

### DISCUSSION THREAD

**Discussion by hardillb** [emqx/emqx#14191] [Q&A] [answered]
I'm looking for a nice way to present a list of recently used topics to a user.

I'm trying to find something other than having a super user client subscribed to `#` and using that to sample the topics, as this means that client would have to receive all the payloads as well as.

I was initially given hope by the `/api/v5/topics` REST endpoint, but having played with this it appears to only list subscription patterns, not the list of topics for published messages.

While I understand why brokers don't normally collect this information (the overhead of logging on every message and deciding when to remove a topic from the list if no more messages are published).

Does anybody have any better solutions rather than subscribing to `#`?

**Comment by zmstone** [comment, marked as answer]
You can make use of the rule-engine.

1. create a rule like this

```
# this is not a subscription, but a evaluation of a hook callback
select topic as published_topic from "#"
```

3. add a republish action to send a new messages with payload rendered as `${published_topic}` and pick a hard coded new topic name e.g. `topic-names`

4. start a subscriber client which subscribes to `topic-names` topic to receive published-topics in the payload.

Alternative solutions

- make use of the webhook data integration to send the topics to a http server which can collect and deduplicate topics
- make use of the http authorization, the http server will receive all publish requests. the advantage is authorization result is cached, so the http server will receive less requests than using webhook integration
- More options if you choose EMQX enterprise edition, e.g. Kafka, MySQL, Postgres, Clickhouse, etc. 
  - For databases, you can use UPSERT statements, e.g. postgres `ON CONFLICT` clause.
  - if you choose Kafka, you can make use of Kafka compacted topics, and produce Kafka message with `${published_topic}` as the key, so Kafka will deduplicate.

**Reply by hardillb** [reply]
I really like the HTTP Auth option as that is already in play
